src/audit/cache_poisoning.rs

```diff
@@ -0,0 +1,282 @@
+            uses: Uses::from_step("ruby/setup-ruby").unwrap(),
+            cache_control: CacheControl::OptIn("bundler-cache"),
+            control_value: ControlValue::Boolean,
+            caching_by_default: false,
+        },
+        // https://github.com/PyO3/maturin-action/blob/main/action.yml
+        CacheAwareAction {
+            uses: Uses::from_step("PyO3/maturin-action").unwrap(),
+            cache_control: CacheControl::OptIn("sccache"),
+            control_value: ControlValue::Boolean,
+            caching_by_default: false,
+        },
+    ]
+});
+
+#[derive(PartialEq)]
+enum CacheUsage {
+    ConditionalOptIn,
+    DirectOptIn,
+    DefaultActionBehaviour,
+}
+
+pub(crate) struct CachePoisoning;
+
+audit_meta!(
+    CachePoisoning,
+    "cache-poisoning",
+    "runtime artifacts potentially vulnerable to a cache poisoning attack"
+);
+
+impl CachePoisoning {
+    fn trigger_used_when_publishing_artifacts(&self, trigger: &Trigger) -> bool {
+        match trigger {
+            Trigger::BareEvent(event) => *event == BareEvent::Release,
+            Trigger::BareEvents(events) => events.contains(&BareEvent::Release),
+            Trigger::Events(events) => match &events.push {
+                OptionalBody::Body(body) => body.tag_filters.is_some(),
+                _ => false,
+            },
+        }
+    }
+
+    fn evaluate_default_action_behaviour(action: &CacheAwareAction) -> Option<CacheUsage> {
+        if action.caching_by_default {
+            Some(CacheUsage::DefaultActionBehaviour)
+        } else {
+            None
+        }
+    }
+
+    fn evaluate_user_defined_opt_in(
+        cache_control_input: &str,
+        env: &Env,
+        action: &CacheAwareAction,
+    ) -> Option<CacheUsage> {
+        match env.get(cache_control_input) {
+            None => None,
+            Some(EnvValue::Boolean(value)) => {
+                if action.control_value == ControlValue::Boolean {
+                    if *value {
+                        Some(CacheUsage::DirectOptIn)
+                    } else {
+                        // Explicitly opts out from caching
+                        None
+                    }
+                } else {
+                    None
+                }
+            }
+            Some(EnvValue::Number(_)) => {
+                // Not sure about what to do here!
+                None
+            }
+            Some(EnvValue::String(value)) => match ExplicitExpr::from_curly(value) {
+                None => {
+                    if action.control_value == ControlValue::String {
+                        Some(CacheUsage::DirectOptIn)
+                    } else {
+                        None
+                    }
+                }
+                Some(_) => Some(CacheUsage::ConditionalOptIn),
+            },
+        }
+    }
+
+    fn evaluate_cache_usage(&self, uses: &str, env: &Env) -> Option<CacheUsage> {
+        let (reference, _) = uses.split_once("@")?;
```

Comment:
Nitpick: I think ad-hoc parsing here is going to be brittle, e.g. if a user does `uses: ruby/setup-ruby` (with no `@`-pin), which we'd currently end up silently skipping here due to the `split_once` usage.

Instead, maybe we should parse with `Uses::from_common` here?

Reply:
Definetely yes! Check my comment below 🙂